### PR TITLE
Refactor cloud provider

### PR DESCRIPTION
**Pull request by yoheiueda**
This PR introduces a common cloud provider interface, and refactors cloud API adaptor to reduce duplicated code among cloud providers.

### CODE REVIEW

**Comment by yoheiueda** [comment]
This PR is the first step to fix #458 

* Introduce a common package for cloud providers (fixes #459)
* Refactor ibmcloud provide code (fixes #460)

File: .gitignore [hunk 1]
```diff
@@ -8,8 +8,8 @@ umoci
 skopeo
 
 # Binaries
-agent-protocol-forwarder
-cloud-api-adaptor
+/agent-protocol-forwarder
```
Comment (jtumber-ibm) [inline .gitignore:11]: Could you revert this change as it will no longer ignore binaries in sub-folders?

File: cmd/cloud-api-adaptor/cloudmgr/manager.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/cloud"
+)
+
+type Cloud struct {
+	ParseCmd    func(flags *flag.FlagSet)
+	LoadEnv     func()
+	NewProvider func() (cloud.Provider, error)
+}
+
+var cloudTable map[string]*Cloud = make(map[string]*Cloud)
```
Comment (jtumber-ibm) [inline cmd/cloud-api-adaptor/cloudmgr/manager.go:19]: As only one provider will be available per binary. Creating a one entry map seems redundant, I'd suggest the following approch instead. Code shorten (let me know if you would like more details)

manager.go
```go
package cloudmgr

type Cloud interface {
	ParseCmd(flags *flag.FlagSet)
	LoadEnv()
	NewProvider() (cloud.Provider, error)
}
```
ibmcloud.go
```go
//go:build ibmcloud
package cloudmgr
type IBMCloudMgr struct{}

func Get() Cloud {
	return IBMCloudMgr{}
}
func (IBMCloudMgr) ParseCmd(flags *flag.FlagSet) { ... }
func (IBMCloudMgr) LoadEnv() { ... }
func (IBMCloudMgr) NewProvider() (cloud.Provider, error) { ... }
```

Likewise following providers would follow pattern

aws.go
```go
//go:build aws
package cloudmgr
type AWSCloudMgr struct{}
func Get() Cloud {
	return AWSCloudMgr{}
}
etc...
```

File: .gitignore (hunk 1 above)
Comment (yoheiueda) [reply .gitignore:11]: OK. I'll revert his.

File: pkg/adaptor/cloud/cloud.go [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+	}
+
+	s.sandboxes[sid] = sandbox
+
+	return nil
+}
+
+func (s *cloudService) getSandbox(id sandboxID) (*sandbox, error) {
+
+	sid := sandboxID(id)
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go:93]: I'll fix it.

File: pkg/adaptor/cloud/cloud.go [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+			return "", fmt.Errorf("failed to get instance ID: %w", ctx.Err())
+		default:
+		}
+
+		s.cond.Wait()
+	}
+}
+
+func (s *cloudService) Version(ctx context.Context, req *pb.VersionRequest) (*pb.VersionResponse, error) {
+	return &pb.VersionResponse{Version: Version}, nil
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go:177]: The `Version` method is an implementation of this protobuf API call. 

https://github.com/kata-containers/kata-containers/blob/f23b087145ed48ed9e91e5512a2abd3b7b2d649f/src/runtime/protocols/hypervisor/hypervisor.proto#L12

The golang version of the API call is defined here, and contains error as return type.

https://github.com/kata-containers/kata-containers/blob/f23b087145ed48ed9e91e5512a2abd3b7b2d649f/src/runtime/protocols/hypervisor/hypervisor.pb.go#L929

File: cmd/cloud-api-adaptor/cloudmgr/manager.go (hunk 2 above)
Comment (yoheiueda) [reply cmd/cloud-api-adaptor/cloudmgr/manager.go:19]: Thank you for your comment. Let me explain my intention here.

My intention is to add ability to support multiple providers in a single binary. We can still switch cloud provider by specifying a cloud provider name in a first command line argument, as we do in the current implementation.

When we build a binary with multiple build tags like `go build -tags ibmcloud,aws ...`, the output binary will support both ibmcloud and aws.  When we build a binary with a single build tag like `go build -tags ibmcloud ...`, the output binary will support ibmcloud only.

Comment (jtumber-ibm) [reply cmd/cloud-api-adaptor/cloudmgr/manager.go:19]: I didn't think we'd support running multiple providers from the same binary? I'm not sure I can see a scenario where that would be used.

Comment (bpradipt) [reply cmd/cloud-api-adaptor/cloudmgr/manager.go:19]: How about discussing the pros/cons of supporting multiple providers in a single binary via a separate github issue and if required in the community interlock ?

**Comment by bpradipt** [comment]
@yoheiueda can you briefly summarise the new code layout, interfaces etc? This will be very helpful.

File: cmd/cloud-api-adaptor/cloudmgr/manager.go (hunk 2 above)
Comment (yoheiueda) [reply cmd/cloud-api-adaptor/cloudmgr/manager.go:19]: OK. I'll raise a new issue for discussing this point.

**Comment by yoheiueda** [comment]
- pkg/adaptor/cloud

This package contains the common interface `Provider` in the `cloud` package.

Each cloud provider's implemention of this interface will be located under `pkg/adaptor/cloud/` like this.

- pkg/adaptor/cloud/ibmclod
- pkg/adaptor/cloud/aws
- ...

The `Provider` interface only has the following two methods for now.

https://github.com/confidential-containers/cloud-api-adaptor/blob/b750f9bc3b8891d8ffe6014d5e92ad11d0b50967/pkg/adaptor/cloud/cloud.go#L36-L39

The `CreateInstance` returns a `Instance` object defined here. Note that this object contains IP address information of the newly created instance.

https://github.com/confidential-containers/cloud-api-adaptor/blob/b750f9bc3b8891d8ffe6014d5e92ad11d0b50967/pkg/adaptor/cloud/cloud.go#L41-L45

The handler code of TTRPC calls are refactored to contain only cloud-neutral code by using the new `Provider` interface.
https://github.com/confidential-containers/cloud-api-adaptor/blob/b750f9bc3b8891d8ffe6014d5e92ad11d0b50967/pkg/adaptor/cloud/cloud.go#L180-L367


The implementation of the `Provider` interface for ibmcloud is found here. 
https://github.com/confidential-containers/cloud-api-adaptor/blob/b750f9bc3b8891d8ffe6014d5e92ad11d0b50967/pkg/adaptor/cloud/ibmcloud/provider.go

The main function now only contains a common cloud-platform-neutral code.
https://github.com/confidential-containers/cloud-api-adaptor/blob/b750f9bc3b8891d8ffe6014d5e92ad11d0b50967/cmd/cloud-api-adaptor/main.go#L58-L90

The other cloud-platform-specific code has been move into the `cloudmgr` package at [`/cmd/cloud-api-adaptor/cloudmgr`](https://github.com/confidential-containers/cloud-api-adaptor/tree/b750f9bc3b8891d8ffe6014d5e92ad11d0b50967/cmd/cloud-api-adaptor/cloudmgr)

The files for the cloud-platform-specific code have build tags to selectively build cloud provider code.

File: pkg/adaptor/cloud/cloud.go [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+			logger.Print(err)
+		}
+	}()
+
+	sid := sandboxID(req.Id)
+
+	if sid == "" {
+		err := errors.New("empty sandbox id")
+		logger.Print(err)
+		return nil, err
```
Comment (bpradipt) [inline pkg/adaptor/cloud/cloud.go:193]: nit: may be having a consistent error reporting will be helpful. Eg using `return nil, fmt.Errorf("empty sandbox id")`

File: pkg/adaptor/cloud/cloud.go [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+		if !(('a' <= c && c <= 'z') || ('0' <= c && c <= '9') || c == '-') {
+			c = '-'
+		}
+		output += string(c)
+	}
+
+	return output
+}
+
+var podvmNamePrefix = "podvm"
```
Comment (bpradipt) [inline pkg/adaptor/cloud/cloud.go:401]: how about specifying this as a const in the beginning itself ?

File: pkg/adaptor/cloud/cloud.go [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+	// cri-o stores the sandbox name in the form of k8s_<pod name>_<namespace>_<uid>_0
+	// Extract the pod name from it.
+	if tmp := strings.Split(sandboxName, "_"); len(tmp) > 1 && tmp[0] == "k8s" {
+		return tmp[1]
+	}
+
+	return sandboxName
+}
+
+func getPodNamespace(annotations map[string]string) string {
```
Comment (bpradipt) [inline pkg/adaptor/cloud/cloud.go:382]: would it make sense to have these helper methods under util ?

File: pkg/adaptor/cloud/cloud.go [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+		output += string(c)
+	}
+
+	return output
+}
+
+var podvmNamePrefix = "podvm"
+
+func GenerateInstanceName(podName, sandboxID string, podvmNameMax int) string {
+
```
Comment (bpradipt) [inline pkg/adaptor/cloud/cloud.go:404]: should this be a helper method under util package ?

**Comment by bpradipt** [comment]
@yoheiueda overall this looks good to me in the first pass. Added some minor comments. 
Also please change the copyright statement to the CNCF ones.
I'll do another pass this week.

File: pkg/adaptor/cloud/cloud.go [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+	podName = sanitize(podName)
+	sandboxID = sanitize(sandboxID)
+
+	podNameLen := len(podName)
+	if podvmNameMax > 0 && len(podvmNamePrefix)+podNameLen+10 > podvmNameMax {
```
Comment (knrt10) [inline pkg/adaptor/cloud/cloud.go:409]: Should we also have `len(podvmNamePrefix)` as a constant?

File: pkg/adaptor/proxy/factory.go [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+		CriSocketPath: criSocketPath,
+	}
+}
+
+func (f *factory) New(socketPath string) AgentProxy {
+
+	agentProxy := NewAgentProxy(socketPath, f.CriSocketPath, f.PauseImage)
+
+	return agentProxy
+}
```
Comment (knrt10) [inline pkg/adaptor/proxy/factory.go:27]: How about just?

```suggestion
func (f *factory) New(socketPath string) AgentProxy {
	return NewAgentProxy(socketPath, f.CriSocketPath, f.PauseImage)
}
```

File: pkg/adaptor/server.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+	s := &server{
+		socketPath:    cfg.SocketPath,
+		cloudService:  cloudService,
+		vmInfoService: vmInfoService,
+		workerNode:    workerNode,
+		readyCh:       make(chan struct{}),
+		stopCh:        make(chan struct{}),
+	}
+
+	return s
```
Comment (knrt10) [inline pkg/adaptor/server.go:78]: Since `s` is not been used furthermore. How about just?

```suggestion
return &server{
		socketPath:    cfg.SocketPath,
		cloudService:  cloudService,
		vmInfoService: vmInfoService,
		workerNode:    workerNode,
		readyCh:       make(chan struct{}),
		stopCh:        make(chan struct{}),
	}
```

File: pkg/adaptor/vminfo/service.go [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+func NewService(cloudService cloud.Service) pb.PodVMInfoService {
+	return &podVMInfoService{cloudService}
+}
+
+func (s *podVMInfoService) GetInfo(ctx context.Context, req *pb.GetInfoRequest) (*pb.GetInfoResponse, error) {
+
+	instanceID, err := s.cloudService.GetInstanceID(ctx, req.PodNamespace, req.PodName, req.Wait)
+	if err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			return nil, status.Errorf(codes.DeadlineExceeded, "failed to get VM ID for %s:%s: %s", req.PodNamespace, req.PodName, err.Error())
```
Comment (knrt10) [inline pkg/adaptor/vminfo/service.go:26]: redundant `failed` keyword in an error. Can we please remove it?

**Review by knrt10 — CHANGES_REQUESTED** [review]
Awesome work @yoheiueda 🎉 . Sorry for NIT picking, requested some of the changes for `failed` keywords in error.

File: pkg/adaptor/cloud/cloud.go (hunk 8 above)
Comment (knrt10) [reply pkg/adaptor/cloud/cloud.go:404]: +1 for this.

File: pkg/adaptor/proxy/factory.go [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+func NewFactory(pauseImage, criSocketPath string) Factory {
+	return &factory{
+		PauseImage:    pauseImage,
+		CriSocketPath: criSocketPath,
+	}
+}
+
+func (f *factory) New(socketPath string) AgentProxy {
```
Comment (surajssd) [inline pkg/adaptor/proxy/factory.go:22]: I think this should be named as `NewAgentProxy` because with plain `New` it signifies that we are creating new of the same object we are referring to. Which is kind of misleading.

And the actual `NewAgentProxy` can be made private because the individual cloud providers won't be calling this function any more. Infact the whole code can sit here.

File: pkg/adaptor/server.go [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/cloud"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/proxy"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/vminfo"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/podnetwork"
+
+	"github.com/containerd/ttrpc"
+	pbHypervisor "github.com/kata-containers/kata-containers/src/runtime/protocols/hypervisor"
+
+	pbPodVMInfo "github.com/confidential-containers/cloud-api-adaptor/proto/podvminfo"
```
Comment (surajssd) [inline pkg/adaptor/server.go:22]: nit: This should be grouped with the other `github.com/confidential-containers/cloud-api-adaptor` imports right?

File: pkg/adaptor/cloud/cloud.go (hunk 5 above)
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/cloud/cloud.go [hunk 15]
```diff
@@ -0,0 +1,420 @@
...
+			}
+		}
+
+		if !wait {
+			return "", nil
+		}
+
+		select {
+		case <-ctx.Done():
+			return "", fmt.Errorf("failed to get instance ID: %w", ctx.Err())
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/cloud/cloud.go [hunk 16]
```diff
@@ -0,0 +1,420 @@
...
+	podNetworkConfig, err := s.workerNode.Inspect(netNSPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to inspect netns %s: %w", netNSPath, err)
+	}
+
+	daemonConfig := forwarder.Config{
+		PodNamespace: namespace,
+		PodName:      pod,
+		PodNetwork:   podNetworkConfig,
+	}
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go:216]: I agree. 

forwarder.Config is reference from cloud providers other than ibmcloud. In this PR, I only refactored ibmcloud, and keep the other cloud providers unchanged.

For now, I changed the definition of `sandbox` instead. We will change the definition of forwader.Config, when we refactored all of the cloud providers.

File: pkg/adaptor/cloud/cloud.go [hunk 17]
```diff
@@ -0,0 +1,420 @@
...
+	daemonJSON, err := json.MarshalIndent(daemonConfig, "", "    ")
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate JSON data: %w", err)
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/cloud/cloud.go [hunk 18]
```diff
@@ -0,0 +1,420 @@
...
+	}
+
+	// Store daemon.json in worker node for debugging
+	podDir := filepath.Join(s.podsDir, string(sid))
+	if err := os.MkdirAll(podDir, os.ModePerm); err != nil {
+		return nil, fmt.Errorf("failed to create a directory for daemon.json: %s, %w", podDir, err)
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/cloud/cloud.go [hunk 19]
```diff
@@ -0,0 +1,420 @@
...
+		if err != nil {
+			logger.Print(err)
+		}
+	}()
+
+	sid := sandboxID(req.Id)
+
+	sandbox, err := s.getSandbox(sid)
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/cloud/cloud.go [hunk 20]
```diff
@@ -0,0 +1,420 @@
...
+		pod:              pod,
+		namespace:        namespace,
+		netNSPath:        netNSPath,
+		agentProxy:       agentProxy,
+		podNetworkConfig: podNetworkConfig,
+		cloudConfig:      cloudConfig,
+	}
+
+	if err := s.addSandbox(sid, sandbox); err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/cloud/cloud.go [hunk 21]
```diff
@@ -0,0 +1,420 @@
...
+	}
+
+	instance, err := s.provider.CreateInstance(ctx, sandbox.pod, string(sid), sandbox.cloudConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create an instance : %w", err)
+	}
+
+	if err := s.setInstance(sid, instance.ID, instance.Name); err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/cloud/cloud.go [hunk 22]
```diff
@@ -0,0 +1,420 @@
...
+	return &pb.StartVMResponse{}, nil
+}
+
+func (s *cloudService) StopVM(ctx context.Context, req *pb.StopVMRequest) (*pb.StopVMResponse, error) {
+
+	sid := sandboxID(req.Id)
+
+	sandbox, err := s.getSandbox(sid)
+	if err != nil {
+		err = fmt.Errorf("failed to stop VM: %v", err)
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/cloud/cloud.go [hunk 23]
```diff
@@ -0,0 +1,420 @@
...
+	}
+
+	// Store daemon.json in worker node for debugging
+	podDir := filepath.Join(s.podsDir, string(sid))
+	if err := os.MkdirAll(podDir, os.ModePerm); err != nil {
+		return nil, fmt.Errorf("failed to create a directory for daemon.json: %s, %w", podDir, err)
+	}
+	daemonJSONPath := filepath.Join(podDir, "daemon.json")
+	if err := os.WriteFile(daemonJSONPath, daemonJSON, 0666); err != nil {
+		return nil, fmt.Errorf("failed to store %s: %w", daemonJSONPath, err)
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/cloud/cloud.go [hunk 24]
```diff
@@ -0,0 +1,420 @@
...
+	sid := sandboxID(req.Id)
+
+	sandbox, err := s.getSandbox(sid)
+	if err != nil {
+		return nil, err
+	}
+
+	instance, err := s.provider.CreateInstance(ctx, sandbox.pod, string(sid), sandbox.cloudConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create an instance : %w", err)
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/cloud/cloud.go [hunk 25]
```diff
@@ -0,0 +1,420 @@
...
+	}
+
+	if err := s.setInstance(sid, instance.ID, instance.Name); err != nil {
+		return nil, err
+	}
+
+	logger.Printf("created an instance %s for sandbox %s", instance.Name, sid)
+
+	if err := s.workerNode.Setup(sandbox.netNSPath, instance.IPs, sandbox.podNetworkConfig); err != nil {
+		return nil, fmt.Errorf("failed to set up pod network tunnel on netns %s: %w", sandbox.netNSPath, err)
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/cloud/cloud.go (hunk 9 above)
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/proxy/factory.go (hunk 10 above)
Comment (yoheiueda) [reply pkg/adaptor/proxy/factory.go]: done

File: pkg/adaptor/server.go (hunk 11 above)
Comment (yoheiueda) [reply pkg/adaptor/server.go]: done

File: cmd/cloud-api-adaptor/cloudmgr/manager.go [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+
+func Get(name string) *Cloud {
+	return cloudTable[name]
+}
+
+func List(name string) []string {
```
[the remark this answers is not in the document]
Comment (yoheiueda) [reply cmd/cloud-api-adaptor/cloudmgr/manager.go]: This is related to the support of multiple cloud providers. I'll delete this for now.

File: pkg/adaptor/proxy/factory.go (hunk 13 above)
Comment (yoheiueda) [reply pkg/adaptor/proxy/factory.go:22]: I agree.

NewAgentProxy is reference from cloud providers other than ibmcloud.   is reference from cloud providers other than ibmcloud. In this PR, I only refactored ibmcloud, and keep the other cloud providers unchanged.

I'll change the names of these functions, when we refactored all of the cloud providers.

File: pkg/adaptor/server.go (hunk 14 above)
Comment (yoheiueda) [reply pkg/adaptor/server.go:21]: done

File: .gitignore (hunk 1 above)
Comment (yoheiueda) [reply .gitignore]: done

File: pkg/adaptor/cloud/cloud.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: cmd/cloud-api-adaptor/cloudmgr/manager.go (hunk 2 above)
Comment (yoheiueda) [reply cmd/cloud-api-adaptor/cloudmgr/manager.go]: I have changed the struct to interface, but I couldn't remove the `cloudTable` map. I am currently refactoring the ibmcloud provider code only, and not touching the other providers. 

If I define `Get()` with the `ibmcloud` build tag, CI tests for the other fail for now, since I don't prepare `Get()` methods in the other providers.

I'll change this to the proposed code, when all of the cloud providers are refactored.

File: pkg/adaptor/vminfo/service.go (hunk 12 above)
Comment (yoheiueda) [reply pkg/adaptor/vminfo/service.go]: done

File: pkg/adaptor/cloud/cloud.go (hunk 6 above)
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: done

File: pkg/adaptor/cloud/cloud.go (hunk 7 above)
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go:379]: This function is private, and only used in this package. Should we move this function to the util package?

File: pkg/adaptor/cloud/cloud.go (hunk 8 above)
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go:399]: This function is intended to used in cloud provider code under `pkg/adaptor/cloud/*`.  Should we move this function to the `pkg/util` package?

File: pkg/adaptor/cloud/cloud.go (hunk 7 above)
Comment (bpradipt) [reply pkg/adaptor/cloud/cloud.go:379]: I was thinking if we move this function to util and make it public it'll be useful in general -e.g for csi wrapper or other stuffs. 
Let me know what you think?

File: pkg/adaptor/cloud/cloud.go (hunk 8 above)
Comment (bpradipt) [reply pkg/adaptor/cloud/cloud.go:399]: same comment as above - https://github.com/confidential-containers/cloud-api-adaptor/pull/329#discussion_r1083852143

**Comment by stevenhorsman** [comment]
@knrt10 - Hey Kautilya - I just wanted to check if you are fine with Yohei's updates and responses to your comments, or if there was anything outstanding you were waiting for before updating your review? Thanks!

**Comment by knrt10** [comment]
Aah, forgot to update my review. Thanks for addressing the changes. LGTM.

**Review by knrt10 — APPROVED** [review]
Aah, forgot to update my review. Thanks for addressing the changes. LGTM.

File: cmd/cloud-api-adaptor/cloudmgr/manager.go (hunk 2 above)
Comment (yoheiueda) [reply cmd/cloud-api-adaptor/cloudmgr/manager.go]: I tried to remove the `cloudTable` map, and suppress golint errors by editing `.golangci.yaml`, but I can suppress the following errors.

```
golangci-lint (1.18): cmd/cloud-api-adaptor/cloudmgr/azure.go#L8
`Get` redeclared in this block (typecheck)
golangci-lint (1.18): cmd/cloud-api-adaptor/cloudmgr/aws.go#L8
other declaration of Get (typecheck)
golangci-lint (1.18): cmd/cloud-api-adaptor/cloudmgr/ibmcloud.go#L15
`Get` redeclared in this block (typecheck)
golangci-lint (1.18): cmd/cloud-api-adaptor/cloudmgr/libvirt.go#L8
`Get` redeclared in this block (typecheck)
```

For now, I continue to use  the `cloudTable` map, and I will come back here when the other cloud provider refactoring is done.

File: pkg/adaptor/cloud/cloud.go (hunk 7 above)
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: OK. I move the function to the `util` package.

File: pkg/adaptor/cloud/cloud.go (hunk 8 above)
Comment (yoheiueda) [reply pkg/adaptor/cloud/cloud.go]: OK. I move the function to the `util` package.

**Comment by yoheiueda** [comment]
@knrt10  thank you for your review. 

@bpradipt I have updated the code based on your comments. Could you check again?

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @yoheiueda

File: cmd/cloud-api-adaptor/cloudmgr/manager.go (hunk 2 above)
Comment (yoheiueda) [reply cmd/cloud-api-adaptor/cloudmgr/manager.go]: I raise an issue for this topic.
 #573
